usaspending_api/etl/elasticsearch_loader_helpers/utilities.py

```diff
@@ -2,69 +2,53 @@
-from typing import Optional
+from dataclasses import dataclass
+from django.conf import settings
+from pathlib import Path
+from random import choice
+from typing import Optional, List
+
 
 from usaspending_api.common.helpers.sql_helpers import get_database_dsn_string
 
 logger = logging.getLogger("script")
 
 
-class DataJob:
-    def __init__(self, *args):
-        self.name = args[0]
-        self.index = args[1]
-        self.fy = args[2]
-        self.csv = args[3]
-        self.count = None
+@dataclass
+class WorkerNode:
+    """Contains details for a worker node to perform micro ETL step"""
+
+    name: str
+    index: str
+    sql: str
+    primary_key: str
+    is_incremental: bool
+    transform_func: callable = None
+    # ids: List[int] = field(default_factory=list)
 
 
 def chunks(l, n):
-    """Yield successive n-sized chunks from l."""
+    """Yield successive n-sized chunks from l"""
     for i in range(0, len(l), n):
         yield l[i : i + n]
 
 
-def convert_postgres_array_as_string_to_list(array_as_string: str) -> Optional[list]:
-    """
-        Postgres arrays are stored in CSVs as strings. Elasticsearch is able to handle lists of items, but needs to
-        be passed a list instead of a string. In the case of an empty array, return null.
-        For example, "{this,is,a,postgres,array}" -> ["this", "is", "a", "postgres", "array"].
-    """
-    return array_as_string[1:-1].split(",") if len(array_as_string) > 2 else None
-
-
-def convert_postgres_json_array_as_string_to_list(json_array_as_string: str) -> Optional[dict]:
+def convert_postgres_json_array_to_list(json_array: dict) -> Optional[List]:
     """
         Postgres JSON arrays (jsonb) are stored in CSVs as strings. Since we want to avoid nested types
         in Elasticsearch the JSON arrays are converted to dictionaries to make parsing easier and then
         converted back into a formatted string.
     """
-    if json_array_as_string is None or len(json_array_as_string) == 0:
+    if json_array is None or len(json_array) == 0:
         return None
     result = []
-    json_array = json.loads(json_array_as_string)
     for j in json_array:
         for key, value in j.items():
             j[key] = "" if value is None else str(j[key])
         result.append(json.dumps(j, sort_keys=True))
     return result
 
 
-def process_guarddog(process_list):
-    """
-        pass in a list of multiprocess Process objects.
-        If one errored then terminate the others and return True
-    """
-    for proc in process_list:
-        # If exitcode is None, process is still running. exit code 0 is normal
-        if proc.exitcode not in (None, 0):
-            msg = f"Script proccess failed!!! {proc.name} exited with error {proc.exitcode}. Terminating all processes."
-            logger.error(format_log(msg))
-            [x.terminate() for x in process_list]
-            return True
-    return False
-
-
 def execute_sql_statement(cmd, results=False, verbose=False):
```

Comment:
@kwhickey How concerned are you with memory usage?

The test run ops yesterday was a stress test. After doing the math, I see
Total box memory avail: 15.5GB (15,872MB)
Number of sub-processes: 50
Completely ignoring everything else on the box, _if we assume_ the script consumed 100% of available memory, then 51 total python processes (parent + 50 sub-processes) used on average 311MB per process. (Assuming the parent process consumes the same amount, which was more or less accurate during local testing.

Each sub-process contained ~12K rows, ignoring any process or script overhead, each record of X columns of various datatypes consumed 26KB. That's looking at the very worse-case math. I have no insights on the overhead so this memory per record is either slightly or greatly smaller. Does that sound unreasonably high?

Also, there is no reason to use OrderedDict in Python3.7. That is only necessary if the code supports older Python versions (Python3.5 and earlier). That isn't a Python 3.7 thing but the library dev adding the overhead no longer necessary since the default `dict` type is ordered and very memory efficient.

Reply:
Given we may have 3 of these running at the same time (awards, transactions, accounts) we should get memory-per-process as low as we an to allow more processes to run in parallel, and increase throughput. 

300+ MB per proc I consider pretty high for moving just 12k database rows. 12k records in uncompressed CSV format from UTM is 11.2MB on disk, ~0.96 KB per record.

Maybe not in version 1 implementation but we need to try out different things to get it lower. At 300+ MB per proc, something is being wasteful or needing to clean up after itself.

Reply:
@kwhickey comparing the CSV rows to the python record isn't apples-to-apples for many reasons. Firstly the CSV doesn't contain all of the data fields, nor all of the values, and the different datatypes aren't in the CSV. Whatever data object contains the data in Python will include overhead of some kind that CSV intentionally omits.

The baseline memory usage on the jenkins box is ~1.86GB. Assuming the script test run used 100% of available memory, it calculates to about 273MB per Python process. I don't have evidence but I assume the record data is a fraction of that.

In the previous comment I was showing an unrealistic high water mark since memory is consumed by other objects such as the Elasticsearch client, and psycopg client, the python interpreter.... I was hoping you would realize that 13GB is tiny for a monolith ETL worker node. I would be curious to see a Spark cluster run on only 13GB.

Reply:
It's hard to get a sense of how much memory a single spark executor uses along the way. There's a max set at I think 512 MB that it cannot exceed. Then each partition is split into blocks that may get distributed around the pool of executors.

Still, less memory means more processes and faster throughput, so we eventually need to test out ways to reduce it to the minimum.